modules/basics/src/main/java/com/opengamma/strata/basics/market/ExtendedMarketData.java

```diff
@@ -0,0 +1,457 @@
+/**
+ * Copyright (C) 2015 - present by OpenGamma Inc. and the OpenGamma group of companies
+ *
+ * Please see distribution for license.
+ */
+package com.opengamma.strata.basics.market;
+
+import java.io.Serializable;
+import java.time.LocalDate;
+import java.util.Map;
+import java.util.NoSuchElementException;
+import java.util.Optional;
+import java.util.Set;
+
+import org.joda.beans.Bean;
+import org.joda.beans.BeanBuilder;
+import org.joda.beans.BeanDefinition;
+import org.joda.beans.ImmutableBean;
+import org.joda.beans.JodaBeanUtils;
+import org.joda.beans.MetaProperty;
+import org.joda.beans.Property;
+import org.joda.beans.PropertyDefinition;
+import org.joda.beans.impl.direct.DirectFieldsBeanBuilder;
+import org.joda.beans.impl.direct.DirectMetaBean;
+import org.joda.beans.impl.direct.DirectMetaProperty;
+import org.joda.beans.impl.direct.DirectMetaPropertyMap;
+
+import com.opengamma.strata.collect.timeseries.LocalDateDoubleTimeSeries;
+
+/**
+ * A set of market data where an item has been added or overridden.
+ * <p>
+ * This decorates an underlying instance to add or replace a single identifier.
+ */
+@BeanDefinition(builderScope = "private", constructorScope = "package")
+final class ExtendedMarketData<T>
```

Comment:
Are these going to end up deeply nested? I can see that becoming a pain when debugging.

If so it might be worth having another implementation backed by a map of overrides to keep things simpler. This class could override `withValue` to create a map-backed version and the map-backed version could override `withValue` to create another map-backed impl.

Having said that, those changes can be made later.